False positive in unneeded_override

### New Issue Checklist

- [x] Updated SwiftLint to the latest version
- [x] I searched for [existing GitHub issues](https://github.com/realm/SwiftLint/issues)

### Describe the bug

In the following code, the override is deemed redundant, even though it provides an default parameter:

```swift
class Foo {
    func bar(value: String) {
        print(value)
    }
}

class Baz: Foo {
    override func bar(value: String = "Hello") {
        super.bar(value: value)
    }
}
```

##### Complete output when running SwiftLint, including the stack trace and command used

```bash
$ swiftlint lint test.swift
Linting Swift files at paths test.swift
Linting 'test.swift' (1/1)
[snipped]/test.swift:8:5: warning: Unneeded Overridden Functions Violation: Remove overridden functions that don't do anything except call their super (unneeded_override)
Done linting! Found 1 violation, 0 serious in 1 file.
```

### Environment

* SwiftLint version: 0.54.0
* Installation method used: CocoaPods
* Paste your configuration file: N/A
* Are you using [nested configurations](https://github.com/realm/SwiftLint#nested-configurations): No
* Which Xcode version are you using (check `xcodebuild -version`): Xcode 15.0.1 Build version 15A507
* Do you have a sample that shows the issue? See above

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -19,6 +19,11 @@
 
 #### Bug Fixes
 
+* Ignore overridden functions with default parameters in the `unneeded_override`
+  rule as they might change behavior.  
+  [SimplyDanny](https://github.com/SimplyDanny)
+  [#5355](https://github.com/realm/SwiftLint/pull/5355)
+
 * Trigger `nsobject_prefer_isequal` and `redundant_self_in_closure` even in case
   the surrounding declaration is nested in an extension.  
   [SimplyDanny](https://github.com/SimplyDanny)

--- a/Source/SwiftLintBuiltInRules/Rules/Lint/UnneededOverrideRule.swift
+++ b/Source/SwiftLintBuiltInRules/Rules/Lint/UnneededOverrideRule.swift
@@ -20,7 +20,7 @@ private func simplify(_ node: some SyntaxProtocol) -> (any ExprSyntaxProtocol)? 
     if let expr = node.as(AwaitExprSyntax.self) {
         return expr.expression
     } else if let expr = node.as(TryExprSyntax.self) {
-        // Assume using try! / try? changes behavior
+        // Assume using try! / try? changes behavior.
         if expr.questionOrExclamationMark != nil {
             return nil
         }
@@ -37,7 +37,7 @@ private func simplify(_ node: some SyntaxProtocol) -> (any ExprSyntaxProtocol)? 
 
 private func extractFunctionCallSyntax(_ node: some SyntaxProtocol) -> FunctionCallExprSyntax? {
     // Extract the function call from other expressions like try / await / return.
-    // If this returns a non-super calling function that will get filtered out later
+    // If this returns a non-super calling function, it will get filtered out later.
     var syntax = simplify(node)
     while let nestedSyntax = syntax {
         if nestedSyntax.as(FunctionCallExprSyntax.self) != nil {
@@ -79,7 +79,7 @@ private func isUnneededOverride(_ node: FunctionDeclSyntax) -> Bool {
         return false
     }
 
-    // Assume having @available changes behavior
+    // Assume having @available changes behavior.
     if node.attributes.contains(attributeNamed: "available") {
         return false
     }
@@ -92,8 +92,14 @@ private func isUnneededOverride(_ node: FunctionDeclSyntax) -> Bool {
         return false
     }
 
-    // Assume any change in arguments passed means behavior was changed
-    let expectedArguments = node.signature.parameterClause.parameters.map {
+    let declParameters = node.signature.parameterClause.parameters
+    if declParameters.contains(where: { $0.defaultValue != nil }) {
+        // Any default parameter might be a change to the super function.
+        return false
+    }
+
+    // Assume any change in arguments passed means behavior was changed.
+    let expectedArguments = declParameters.map {
         ($0.firstName.text == "_" ? "" : $0.firstName.text, $0.secondName?.text ?? $0.firstName.text)
     }
     let actualArguments = call.arguments.map {

--- a/Source/SwiftLintBuiltInRules/Rules/Lint/UnneededOverrideRuleExamples.swift
+++ b/Source/SwiftLintBuiltInRules/Rules/Lint/UnneededOverrideRuleExamples.swift
@@ -105,6 +105,14 @@ struct UnneededOverrideRuleExamples {
                 })
             }
         }
+        """),
+        Example("""
+        class Baz: Foo {
+            // A default argument might be a change
+            override func bar(value: String = "Hello") {
+                super.bar(value: value)
+            }
+        }
         """)
     ]
 

